src/mon/command_multiplexer.h

```diff
@@ -0,0 +1,312 @@
+have two kinds of operations, preprocess operations and prepare operations. 
+
+Installing a command_registry into a subsystem requires patching the preprocess and prepare pipelines. Additionally,
+the environment a user's function (ie. a command) runs in is centralized into a single state structure.
+
+Implementation note:
+	- rather than having a vector backing everything-- requiring expensive lookups for any search operation because
+    keys in each object must be examined-- it would certainly be possible to back this with a boost::multiindex, or
+    map of shared_ptr<>. Left as a future refinement. If this is done, I suggest making that portion of the implementation
+    a type in common that more closely models an STL associative array, and building this around that implementation.
+*/
+
+namespace ceph {
+
+enum class operation_type : int { preprocess, prepare };
+
+template <typename CommandState, typename CommandResult>
+struct command;
+
+template <typename CommandType>
+struct command_registry
+{
+ using command_type = CommandType;
+
+ // The type returned by an operation:
+ using status_type  = std::optional<typename command_type::result_type>;
+
+ private:
+ command_registry(const command_registry&)              = delete;
+ command_registry& operator=(const command_registry&)   = delete;
+
+ public:
+ const std::string name;      // eg. the name of the registry 
+
+ std::vector<std::unique_ptr<command_type>> commands;
+
+ public:
+ command_registry(std::string_view name_)
+  : name(name_)
+ {}
+
+ virtual ~command_registry() = default;
+
+ public:
+ auto find_cmd_for_tag(std::string_view candidate_tag) const
+ {
+    using std::end;
+    using std::begin;
+
+    auto tag_match = [candidate_tag](const auto& command) {
+        auto tags = command->tags();
+        return end(tags) != std::find(begin(tags), end(tags), candidate_tag);
+    };
+
+    return std::find_if(begin(commands), end(commands), tag_match); 
+ }
+
+ // True if a "tag" is found:
+ bool in_registry(std::string_view tag) const
+ {
+    return end(commands) != find_cmd_for_tag(tag);
+ }
+
+ // True if any tag in the "tags" sequence is found:
+ template <typename K, typename A, template <typename, typename> typename SeqT = std::vector>
+ bool in_registry(const SeqT<K, A>& tags) const
+ {
+    return std::end(tags) != std::find_if(std::begin(tags), std::end(tags), 
+                                         [this](const auto& tag) {
+                                           return this->in_registry(tag);
+                                        });
+ }
+
+ void add_command(std::unique_ptr<CommandType> cmd) 
+ {
+    using namespace std;
+
+    auto tags(cmd->tags());
+
+    if (in_registry(tags)) {
+      ostringstream os;
+      os << "already in registry: ";
+      copy(begin(tags), end(tags), experimental::ostream_joiner(os, ','));
+      throw runtime_error(os.str());
+    }
```

Comment:
Yes, I think that would be a good idea (runtime_error is IMO really too general here!). The <system_error> header is not something I've gotten to really study yet (I see there were some well-regarded C++Now! presentations), so it will be good to have the opportunity. 

With respect to the exact error codes, are you thinking we should adopt one of these? ("https://en.cppreference.com/w/cpp/error/errc")
I have some preference for having more specific rather than more general codes (I don't personally like seeing something like -ENOENT for "configuration element not found", *especially* if it can be ignored), but can be talked into going that way. But does -EEXIST (for instance) really make sense here?

Reply:
+1 to this. I'm fine with this direction-- I need to spend some time understanding <system_error> better before I can implement it, but I'll give it a go.

Reply:
I've left the system_error commit separate for the moment so that we can talk about it in isolation a bit more easily-- I suspect we'll want to fiddle with naming and whatnot. It /does/ seem that we can use it as a pretty general system.